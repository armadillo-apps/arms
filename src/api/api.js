import axios from "./axios";
import queryString from "query-string";

export const fetchApartments = async () => {
  const response = await axios.get("/apartments", { withCredentials: true });
  return response.data;
};

export const fetchOccupants = async () => {
  const response = await axios.get("/occupants", { withCredentials: true });
  return response.data;
};

export const fetchUsers = async () => {
  const response = await axios.get("/users", { withCredentials: true });
  return response.data;
};

export const getApartmentProfileHistory = async apartmentId => {
  const response = await axios.get(
    `/stays/apartmentProfileHistory/${apartmentId}`,
    { withCredentials: true }
  );
  return response.data;
};

export const createNewApartment = async input => {
  const response = await axios.post("/apartments", input, {
    withCredentials: true
  });
  return response.data;
};

export const createNewOccupant = async (
  name,
  employeeId,
  gender,
  remarks,
  homeOffice,
  status
) => {
  const requestBody = {
    name,
    employeeId,
    gender,
    remarks,
    homeOffice,
    status
  };
  const response = await axios.post("/occupants", requestBody, {
    withCredentials: true
  });
  return response.data;
};

export const createNewUser = async (name, email, password, role) => {
  const requestBody = {
    name,
    email,
    password,
    role
  };
  const response = await axios.post("/users/new", requestBody, {
    withCredentials: true
  });
  return response.data;
};

export const createStay = async (
  occupantId,
  apartmentId,
  checkInDate,
  checkOutDate,
  leaseId
) => {
  const requestBody = {
    occupantId: occupantId,
    apartmentId: apartmentId,
    checkInDate: checkInDate,
    checkOutDate: checkOutDate,
    leaseId: leaseId
  };
  const response = await axios.post("/stays", requestBody, {
    withCredentials: true
  });
  return response.data;
};

export const removeStay = async stayId => {
  const response = await axios.delete(`/stays/${stayId}`, {
    withCredentials: true
  });
  return response.data;
};

export const removeUser = async userId => {
  const response = await axios.delete(`/users/${userId}`, {
    withCredentials: true
  });
  return response.data;
};

export const updateOccupant = async (
  occupantId,
  name,
  employeeId,
  gender,
  remarks,
  homeOffice,
  status,
  _id
) => {
  const requestBody = {
    _id,
    name,
    employeeId,
    gender,
    remarks,
    homeOffice,
    status
  };

  const response = await axios.put(`/occupants/${occupantId}`, requestBody, {
    withCredentials: true
  });
  return response.data;
};

export const updateApartment = async (
  apartmentId,
  name,
  address,
  bedrooms,
  capacity,
  country,
  status,
  landlord,
  remarks
) => {
  const requestBody = {
    name,
    address,
    bedrooms,
    capacity,
    country,
    status,
    landlord,
    remarks
  };

  const response = await axios.put(`/apartments/${apartmentId}`, requestBody, {
    withCredentials: true
  });
  return response.data;
};

export const fetchStays = async queryParams => {
  const query = queryString.stringify(queryParams);
  const response = await axios.get(`/stays?${query}`, {
    withCredentials: true
  });
  return response.data;
};

export const loginUser = async (email, password) => {
  const loginDetails = { email, password };
  const login = await axios.post("/users/login", loginDetails, {
    withCredentials: true
  });
  return login.data;
};

export const logoutUser = async () => {
  const logout = await axios.post(
    "/users/logout",
    {},
    { withCredentials: true }
  );
  return logout.data;
};
