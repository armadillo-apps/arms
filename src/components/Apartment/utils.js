import moment from "moment";

export const sortApartmentsByStatus = apartmentList => {
  const activeApartments = apartmentList.filter(
    apartment => apartment.status === "active"
  );

  activeApartments.sort((firstApartment, secondApartment) => {
    let apartmentA = moment(new Date(firstApartment.leases[0].leaseEnd));
    let apartmentB = moment(new Date(secondApartment.leases[0].leaseEnd));
    return apartmentA - apartmentB;
  });

  const inactiveApartments = apartmentList.filter(
    apartment => apartment.status === "inactive"
  );
  return activeApartments.concat(inactiveApartments);
};
