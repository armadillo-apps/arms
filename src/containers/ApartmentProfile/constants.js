export const occupantsCardContent = {
  title: "OCCUPANTS",
  headings: ["NAME", "CHECK-IN", "CHECK-OUT", "REMARKS"],
  emptyMessage: "No occupants yet!"
};

export const leaseCardContent = {
  title: "LEASE",
  headings: ["LEASE START", "LEASE END", "RENT PER MONTH"],
  emptyMessage: "No lease yet!"
};
