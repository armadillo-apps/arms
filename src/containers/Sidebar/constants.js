import ApartmentsLogo from "../../assets/apartments.svg";
import OccupantsLogo from "../../assets/occupants.svg";
import ChangePasswordLogo from "../../assets/edit.svg";

export const sidebarLinks = [
  {
    pathname: "/apartments",
    text: "Apartments",
    imgUrl: ApartmentsLogo,
    imgAlt: "apartments"
  },
  {
    pathname: "/occupants",
    text: "Occupants",
    imgUrl: OccupantsLogo,
    imgAlt: "occupants"
  },
  {
    pathname: "/changePassword",
    text: "Change Password",
    imgUrl: ChangePasswordLogo,
    imgAlt: "change password"
  }
];
