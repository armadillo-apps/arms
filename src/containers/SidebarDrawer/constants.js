import ApartmentsLogo from "../../assets/apartments.svg";
import OccupantsLogo from "../../assets/occupants.svg";
import ChangePasswordLogo from "../../assets/edit.svg";

export const PUBLIC_MENU = [
  {
    pathname: "/apartments",
    text: "Apartments",
    imgUrl: ApartmentsLogo,
    imgAlt: "apartments"
  },
  {
    pathname: "/occupants",
    text: "Occupants",
    imgUrl: OccupantsLogo,
    imgAlt: "occupants"
  },
  {
    pathname: "/changePassword",
    text: "Password",
    imgUrl: ChangePasswordLogo,
    imgAlt: "change password"
  }
];
