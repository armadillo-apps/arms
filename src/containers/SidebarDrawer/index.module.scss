.container {
  width: 20vw;
  height: 100vh;
  background: var(--sidebar-bg-color);
  display: flex;
  border-right: 1px;
  transition: 0.1s;
}

.containerCollapsed {
  width: 5vw;
  transition: 0.1s;
}

.linksContainer {
  padding-top: 20px;
  display: flex;
  flex-direction: column;
}

.linksContainer img {
  width: 20px;
  height: 20px;
  background-size: cover;
  margin-right: 10px;
  overflow: visible;
}

.links {
  display: flex;
  width: 100%;
  text-align: left;
  font-size: var(--sidebar-font-size);
  color: var(--sidebar-font-color-inactive);
  cursor: pointer;
  padding: 30px 20px;
  text-decoration: none;
}

.links:hover {
  background: var(--main-bg-color);
}

.active {
  color: var(--sidebar-font-color-active);
  background: var(--main-bg-color);
  transition: 0.8s;
  border-left: 3px solid;
  border-color: #2349cf;
}

.linkTextHidden {
  // display: none;
  visibility: hidden;
}
