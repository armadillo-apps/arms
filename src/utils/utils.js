import moment from "moment";

export const compareStrings = (str1, str2) =>
  str1.toLowerCase().includes(str2.toLowerCase());

export const isEmpty = objectOrArray => {
  return Object.keys(objectOrArray ?? {}).length === 0;
};

export const formatDate = date => moment(new Date(date)).format("D MMM YY");
