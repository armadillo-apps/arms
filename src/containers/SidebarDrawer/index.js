import React, { useState } from "react";
import { NavLink } from "react-router-dom";
import styles from "./index.module.scss";
import { useUserContext } from "../../context/UserContext";
import { LOGOUT_USER } from "../../reducer/userReducer";
import { logoutUser } from "../../api/api";

import UsersLogo from "../../assets/userManagement.svg";
import LogoutLogo from "../../assets/logout.svg";
import { PUBLIC_MENU } from "./constants";

const SidebarDrawer = () => {
  const { state, dispatch } = useUserContext();
  const [visible, setVisible] = useState(false);
  const userRole = state.role;

  const onLogout = async () => {
    dispatch({ type: LOGOUT_USER });
    await logoutUser();
  };

  const handleHover = () => {
    setVisible(!visible);
  };

  return (
    <div
      className={visible ? styles.container : styles.containerCollapsed}
      onMouseEnter={handleHover}
      onMouseLeave={handleHover}
    >
      <div className={styles.linksContainer}>
        <NavLink
          className={styles.links}
          exact
          to="/"
          activeClassName={styles.active}
        >
          {visible ? "ARMS" : "A"}
        </NavLink>
        {PUBLIC_MENU.map((menuItem, index) => (
          <NavLink
            key={index}
            className={styles.links}
            exact
            to={menuItem.pathname}
            activeClassName={styles.active}
            data-testid={`sideBar-${menuItem.imgAlt}`}
          >
            <img src={menuItem.imgUrl} alt={menuItem.imgAlt} />
            <span className={visible ? null : styles.linkTextHidden}>
              {menuItem.text}
            </span>
          </NavLink>
        ))}
        {userRole === "admin" ? (
          <NavLink
            className={styles.links}
            exact
            to="/users"
            activeClassName={styles.active}
            data-testid="sideBar-users"
          >
            <img src={UsersLogo} alt="user management" />
            <span className={visible ? null : styles.linkTextHidden}>User</span>
          </NavLink>
        ) : (
          ""
        )}
        <NavLink className={styles.links} to="/" onClick={onLogout}>
          <img src={LogoutLogo} alt="logout" />
          <span className={visible ? null : styles.linkTextHidden}>Logout</span>
        </NavLink>
      </div>
    </div>
  );
};

export default SidebarDrawer;
