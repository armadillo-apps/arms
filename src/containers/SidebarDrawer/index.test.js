import React from "react";
import { render } from "@testing-library/react";
import { fireEvent } from "@testing-library/dom";
import renderer from "react-test-renderer";
import { mockUserContext } from "../../../test/utils/mockUserContext";
import { PUBLIC_MENU } from "./constants";
import "@testing-library/jest-dom/extend-expect";
import { BrowserRouter as Router } from "react-router-dom";
import SidebarDrawer from "./index";

import * as data from "../../api/api";

const mockLogout = jest.spyOn(data, "logoutUser");

describe("Sidebar", () => {
  describe("Public", () => {
    beforeEach(() => {
      const guestUser = { role: "guest" };
      mockUserContext(guestUser);
    });

    it("renders correctly", () => {
      const tree = renderer
        .create(
          <Router>
            <SidebarDrawer />
          </Router>
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    });

    it("should toggle collapse on hover", () => {
      const { queryByText, getByAltText, getByText } = render(
        <Router>
          <SidebarDrawer />
        </Router>
      );

      expect(queryByText(/change password/i)).not.toBeInTheDocument();

      fireEvent.mouseOver(getByAltText(/change password/));
      fireEvent.mouseEnter(getByAltText(/change password/));
      expect(getByText(/change password/i)).toBeInTheDocument();

      fireEvent.mouseOver(getByAltText(/change password/));
      fireEvent.mouseLeave(getByAltText(/change password/));
      expect(queryByText(/change password/i)).not.toBeInTheDocument();
    });

    it("should render all public menu links", () => {
      const menuAltTexts = PUBLIC_MENU.map(item => item.imgAlt);

      const { getByAltText } = render(
        <Router>
          <SidebarDrawer />
        </Router>
      );

      menuAltTexts.forEach(item => {
        expect(getByAltText(item)).toBeInTheDocument();
      });
    });

    it("should render logout link", () => {
      mockLogout.mockReturnValue("");

      const { getByAltText } = render(
        <Router>
          <SidebarDrawer />
        </Router>
      );

      const logoutLink = getByAltText(/logout/);
      expect(logoutLink).toBeInTheDocument();
      fireEvent.click(logoutLink);
      expect(mockLogout).toHaveBeenCalledTimes(1);
    });
  });

  describe("Admin only", () => {
    it("should render collapsible user management link", () => {
      const adminUser = { role: "admin" };
      mockUserContext(adminUser);
      const { getByAltText, queryByText, getByText } = render(
        <Router>
          <SidebarDrawer />
        </Router>
      );

      expect(getByAltText(/user management/i)).toBeInTheDocument();

      expect(queryByText(/user management/i)).not.toBeInTheDocument();

      fireEvent.mouseOver(getByAltText(/user management/i));
      fireEvent.mouseEnter(getByAltText(/user management/i));
      expect(getByText(/user management/i)).toBeInTheDocument();

      fireEvent.mouseOver(getByAltText(/user management/));
      fireEvent.mouseLeave(getByAltText(/user management/));
      expect(queryByText(/user management/i)).not.toBeInTheDocument();
    });
  });
});
