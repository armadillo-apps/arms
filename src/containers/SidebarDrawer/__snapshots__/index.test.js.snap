// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar Public renders correctly 1`] = `
<div
  className="containerCollapsed"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <div
    className="linksContainer"
  >
    <a
      aria-current="page"
      className="links active"
      href="/"
      onClick={[Function]}
      style={Object {}}
    >
      A
    </a>
    <a
      aria-current={null}
      className="links"
      data-testid="sideBar-apartments"
      href="/apartments"
      onClick={[Function]}
    >
      <img
        alt="apartments"
        src="apartments.svg"
      />
      
    </a>
    <a
      aria-current={null}
      className="links"
      data-testid="sideBar-occupants"
      href="/occupants"
      onClick={[Function]}
    >
      <img
        alt="occupants"
        src="occupants.svg"
      />
      
    </a>
    <a
      aria-current={null}
      className="links"
      data-testid="sideBar-change password"
      href="/changePassword"
      onClick={[Function]}
    >
      <img
        alt="change password"
        src="edit.svg"
      />
      
    </a>
    
    <a
      aria-current="page"
      className="links active"
      href="/"
      onClick={[Function]}
      style={Object {}}
    >
      <img
        alt="logout"
        src="logout.svg"
      />
      
    </a>
  </div>
</div>
`;
